File: controller/position_actuation.py
"""Stateful joint-velocity limiting and position-command integration."""

from dataclasses import dataclass

import numpy as np


def _array(value, name):
    result = np.asarray(value, dtype=float)
    if result.shape != (7,) or not np.all(np.isfinite(result)):
        raise ValueError(f"{name} must be a finite shape-(7,) array")
    copied = result.copy()
    return np.frombuffer(copied.tobytes(), dtype=copied.dtype).reshape(
        copied.shape
    )


@dataclass(frozen=True, slots=True)
class PositionActuationLimits:
    velocity_rad_s: np.ndarray
    lead_rad: float
    lower_position_rad: np.ndarray
    upper_position_rad: np.ndarray

    def __post_init__(self):
        velocity = _array(self.velocity_rad_s, "velocity_rad_s")
        if np.any(velocity <= 0.0):
            raise ValueError("velocity_rad_s must be positive")
        lead = float(self.lead_rad)
        if not np.isfinite(lead) or lead <= 0.0:
            raise ValueError("lead_rad must be finite and positive")
        lower = np.asarray(self.lower_position_rad, dtype=float)
        upper = np.asarray(self.upper_position_rad, dtype=float)
        if lower.shape != (7,) or upper.shape != (7,):
            raise ValueError("position bounds must be shape-(7,) arrays")
        if np.any(np.isnan(lower)) or np.any(np.isnan(upper)):
            raise ValueError("position bounds must not contain NaN")
        if np.any(lower > upper):
            raise ValueError("lower_position_rad must not exceed upper")
        lower = lower.copy()
        upper = upper.copy()
        lower = np.frombuffer(lower.tobytes(), dtype=lower.dtype)
        upper = np.frombuffer(upper.tobytes(), dtype=upper.dtype)
        object.__setattr__(self, "velocity_rad_s", velocity)
        object.__setattr__(self, "lead_rad", lead)
        object.__setattr__(self, "lower_position_rad", lower)
        object.__setattr__(self, "upper_position_rad", upper)


@dataclass(frozen=True, slots=True)
class PositionActuation:
    qdot_speed_clipped: np.ndarray
    qdot_effective: np.ndarray
    command_before_rad: np.ndarray
    command_after_rad: np.ndarray
    speed_saturated: np.ndarray
    lead_clamped: np.ndarray
    range_clamped: np.ndarray

    def __post_init__(self):
        for name in self.__dataclass_fields__:
            value = np.asarray(getattr(self, name)).copy()
            value = np.frombuffer(
                value.tobytes(), dtype=value.dtype
            ).reshape(value.shape)
            object.__setattr__(self, name, value)


class PositionIntegrator:
    """Persistent commanded position; reset only by reconstruction."""

    def __init__(self, measured_position_rad, limits):
        if not isinstance(limits, PositionActuationLimits):
            raise TypeError("limits must be PositionActuationLimits")
        self._limits = limits
        self._command_rad = _array(
            measured_position_rad, "measured_position_rad").copy()

    @property
    def command_rad(self):
        return _array(self._command_rad, "command_rad")

    def step(self, measured_position_rad, requested_velocity_rad_s, dt_s):
        measured = _array(measured_position_rad, "measured_position_rad")
        requested = _array(
            requested_velocity_rad_s, "requested_velocity_rad_s")
        dt = float(dt_s)
        if not np.isfinite(dt) or dt <= 0.0:
            raise ValueError("dt_s must be finite and positive")

        qdot_speed_clipped = np.clip(
            requested,
            -self._limits.velocity_rad_s,
            self._limits.velocity_rad_s,
        )
        speed_saturated = requested != qdot_speed_clipped
        command_before = self._command_rad.copy()
        integrated = command_before + qdot_speed_clipped * dt
        lead_limited = np.clip(
            integrated,
            measured - self._limits.lead_rad,
            measured + self._limits.lead_rad,
        )
        lead_clamped = integrated != lead_limited
        command_after = np.clip(
            lead_limited,
            self._limits.lower_position_rad,
            self._limits.upper_position_rad,
        )
        range_clamped = lead_limited != command_after
        self._command_rad = command_after.copy()
        return PositionActuation(
            qdot_speed_clipped=qdot_speed_clipped,
            qdot_effective=(command_after - command_before) / dt,
            command_before_rad=command_before,
            command_after_rad=command_after,
            speed_saturated=speed_saturated,
            lead_clamped=lead_clamped,
            range_clamped=range_clamped,
        )

File: cpp/src/control/PositionActuation.cpp
#include "control/PositionActuation.h"

#include <cmath>
#include <stdexcept>

namespace srl::control {
namespace {

void RequireFinite(const Vector7& value, const char* name) {
  if (!value.allFinite()) {
    throw std::invalid_argument(std::string(name) +
                                " must be a finite shape-(7,) array");
  }
}

double Clamp(double value, double lower, double upper) {
  return value < lower ? lower : (value > upper ? upper : value);
}

}  // namespace

void PositionActuationLimits::Validate() const {
  RequireFinite(velocity_rad_s, "velocity_rad_s");
  if ((velocity_rad_s.array() <= 0.0).any()) {
    throw std::invalid_argument("velocity_rad_s must be positive");
  }
  if (!std::isfinite(lead_rad) || lead_rad <= 0.0) {
    throw std::invalid_argument("lead_rad must be finite and positive");
  }
  // NaN is rejected; +/-inf bounds are legal (an unlimited actuator).
  if (lower_position_rad.hasNaN() || upper_position_rad.hasNaN()) {
    throw std::invalid_argument("position bounds must not contain NaN");
  }
  if ((lower_position_rad.array() > upper_position_rad.array()).any()) {
    throw std::invalid_argument("lower_position_rad must not exceed upper");
  }
}

PositionIntegrator::PositionIntegrator(const Vector7& measured_position_rad,
                                       const PositionActuationLimits& limits)
    : limits_(limits), command_rad_(measured_position_rad) {
  limits_.Validate();
  RequireFinite(measured_position_rad, "measured_position_rad");
}

void PositionIntegrator::VelocityBounds(const Vector7& measured_position_rad,
                                        double dt_s, Vector7& lower,
                                        Vector7& upper) const {
  RequireFinite(measured_position_rad, "measured_position_rad");
  if (!std::isfinite(dt_s) || dt_s <= 0.0) {
    throw std::invalid_argument("dt_s must be finite and positive");
  }
  for (int index = 0; index < kJoints; ++index) {
    const double command_lower =
        std::max(measured_position_rad(index) - limits_.lead_rad,
                 limits_.lower_position_rad(index));
    const double command_upper =
        std::min(measured_position_rad(index) + limits_.lead_rad,
                 limits_.upper_position_rad(index));
    lower(index) = std::max(-limits_.velocity_rad_s(index),
                            (command_lower - command_rad_(index)) / dt_s);
    upper(index) = std::min(limits_.velocity_rad_s(index),
                            (command_upper - command_rad_(index)) / dt_s);
  }
  if ((lower.array() > upper.array()).any()) {
    throw std::runtime_error(
        "persistent command has no clamp-free velocity interval");
  }
}

PositionActuation PositionIntegrator::Step(const Vector7& measured_position_rad,
                                           const Vector7& requested_velocity_rad_s,
                                           double dt_s) {
  RequireFinite(measured_position_rad, "measured_position_rad");
  RequireFinite(requested_velocity_rad_s, "requested_velocity_rad_s");
  if (!std::isfinite(dt_s) || dt_s <= 0.0) {
    throw std::invalid_argument("dt_s must be finite and positive");
  }

  PositionActuation result;
  result.command_before_rad = command_rad_;
  for (int index = 0; index < kJoints; ++index) {
    const double limit = limits_.velocity_rad_s(index);
    const double clipped =
        Clamp(requested_velocity_rad_s(index), -limit, limit);
    result.qdot_speed_clipped(index) = clipped;
    // Exact inequality, matching NumPy's `requested != qdot_speed_clipped`.
    result.speed_saturated[index] = requested_velocity_rad_s(index) != clipped;

    const double integrated = command_rad_(index) + clipped * dt_s;
    const double lead_limited =
        Clamp(integrated, measured_position_rad(index) - limits_.lead_rad,
              measured_position_rad(index) + limits_.lead_rad);
    result.lead_clamped[index] = integrated != lead_limited;

    const double command_after =
        Clamp(lead_limited, limits_.lower_position_rad(index),
              limits_.upper_position_rad(index));
    result.range_clamped[index] = lead_limited != command_after;

    result.command_after_rad(index) = command_after;
    result.qdot_effective(index) =
        (command_after - result.command_before_rad(index)) / dt_s;
  }
  command_rad_ = result.command_after_rad;
  return result;
}

}  // namespace srl::control
